servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -416,12 +416,13 @@ private Single<C> selectConnection0(final Predicate<C> selector, @Nullable final
                     // Invoke the selector before adding the connection to the pool, otherwise, connection can be
                     // used concurrently and hence a new connection can be rejected by the selector.
                     if (!selector.test(newCnx)) {
-                        // Failure in selection could be temporary, hence add it to the queue and be consistent
-                        // with the fact that select failure does not close a connection.
-                        return newCnx.closeAsync().concat(failed(StacklessConnectionRejectedException.newInstance(
+                        // Failure in selection could be the result of connection factory returning queued connection,
+                        // and not having visibility into max-concurrent-requests. Propagate the exception and rely upon
+                        // retry strategy.
+                        return failed(StacklessConnectionRejectedException.newInstance(
```

Comment:
This is concerning. I understand that right now the only known use-case for the selector to reject a new connection is a described caching use-case. However, we don't know what users do with concurrency controllers (they are public now) and it becomes harder for us to remember about this path if we make any changes for selector in the future. There is a risk to leak a real (not cached) connection. If we return an exception we should add connection to the host.

Reply:
hmmm good call. Currently selector should only reject if it is closed or concurrency exceeded (already added to the host) so I don't think leaking would be and issue now however adding to host is safer.

Note that I will also filter duplicates from the collection to reduce biased toward connections and keep the false-positive selection rate down.